Configuration doc: add rule configuration section (#2291)
sqlfluff/sqlfluff: docs/source/configuration.rst

.. _templateconfig:

Rule Configuration
------------------

Rules can be configured with the `.sqlfluff` config files.

> Common rule configurations can be set in the *[sqlfluff:rules]* section.

Author: In the code the common section is called "generic":

https://github.com/sqlfluff/sqlfluff/blob/7ffea4670c384f575dd7a59ed31da0b93ae8f800/src/sqlfluff/core/rules/base.py#L951-L954

But I'm leaning towards "common" as the user-facing term to use. Please advise :)

Reviewer: It’s called common in the default config too:
https://github.com/sqlfluff/sqlfluff/blob/7ffea4670c384f575dd7a59ed31da0b93ae8f800/src/sqlfluff/core/default_config.cfg#L37-L38

For example:

.. code-block:: cfg

>     [sqlfluff:rules]
>     unquoted_identifiers_policy = all
>     max_line_length = 88

Reviewer: I think it would be worth including all the common ones here to make it more obvious which ones go in this section:

```suggestion
    # Some rules can be configured directly from the config common to other rules.
   [sqlfluff:rules]
   tab_space_size = 4
   max_line_length = 80
   indent_unit = space
   comma_style = trailing
   allow_scalar = True
   single_table_references = consistent
   unquoted_identifiers_policy = all
```

Author: I committed the suggestion, but in what sense are those "all the common ones"?

Reviewer: I got them from the default config: https://github.com/sqlfluff/sqlfluff/blob/main/src/sqlfluff/core/default_config.cfg

As everything needs to be added to that AFAIK I think that is them all

Author: Unfortunately that means that it can be easy to miss adding it here if something new is added. But I suppose we don't promise here that the list would be comprehensive. It's just an example.

Reviewer: Yeah we’ll try to keep it in sync but may will miss something. Still I think it’s good to start with what we know now.

> Rule specific configurations are set in the *[sqlfluff:rules:\*]* sections.

Reviewer: I find that syntax a little confusing to read. What about this:

```suggestion
Rule specific configurations are set in subsections under each rule id.
```

Author: Is it just me or is it confusing to se that the subsections are "under" rule ids? Would this work: "..set in rule specific subsections"?

Reviewer: Yup that works for me.

For example, enforce that keywords are upper case by configuring the rule
:class:`L010 <sqlfluff.core.rules.Rule_L010>`:

.. code-block:: cfg

    [sqlfluff:rules:L010]  # Keywords
    capitalisation_policy = upper

All possible options for rule sections are documented in :ref:`ruleref`.

For an overview of the most common rule configurations that you may want to
tweak, see `Default Configuration`_ (and use :ref:`ruleref` to find the
available alternatives).

> 

Reviewer: Could we add a section on how to disable rules? Both by explaining the `rules` and `exclude_rules` options in the core section as shown in`configuration.html#default-configuration` (maybe with some examples) and also by referencing inline ignore options with a link to `rules.html#inline-ignoring-errors`

:class:`L010 <sqlfluff.core.rules.Rule_L010>`:

.. code-block:: cfg

    [sqlfluff:rules:L010]  # Keywords
    capitalisation_policy = upper

All possible options for rule sections are documented in :ref:`ruleref`.

For an overview of the most common rule configurations that you may want to
tweak, see `Default Configuration`_ (and use :ref:`ruleref` to find the
available alternatives).

Enabling and Disabling Rules
^^^^^^^^^^^^^^^^^^^^^^^^^^^^

To disable individual rules, set :code:`exclude_rules` in the top level section
of sqlfluff configuration. The value is a comma separated list of rule ids.

For example, to disable the rules :class:`L010 <sqlfluff.core.rules.Rule_L022>`
and :class:`L010 <sqlfluff.core.rules.Rule_L027>`:

.. code-block:: cfg

    [sqlfluff]
    exclude_rules = L022, L027

To enable individual rules, configure :code:`rules`, respectively.

For example, to enable :class:`L010 <sqlfluff.core.rules.Rule_L027>`:

.. code-block:: cfg

    [sqlfluff]
    rules = L027

^ **TODO** does this make sense only if overriding an exclusion from a higher level
configuration? Or what does this conf mean? How is it interpreted if

> :code:`exclude_rules = None` but :code:`exclude_rules != None`?

Author: Do you know how this is? If not, I could try to look it up but not feeling super confident :)

Reviewer: Someone looked this up a few days ago and exclude takes priority over rules if that’s what you’re asking.

I think it only makes sense to either to explicitly include some rules, or to include all and explicitly include some of them. Mixing the two doesn’t make sense IMHO.

Author: > exclude takes priority over rules

Oh. So then `rules` only makes sense to be set if `exclude_rules = None`? If `exclude_rules` is set to any value, `rules` will be entirely ignored?

Author: Is there a concept like this though: "overriding an exclusion from a higher level
configuration"? And is it possible to have `.sqlfluff` files in folder hierarchy that overwrite some particular configurations? If yes, then I could imagine `rules` to be set on highest level, but then one could exclude some of those rules with `exclude_rules` in the scope of some subdirectory. Or am I just imagining something that is not even a feature? :)

Reviewer: Ah that’s true. Yes you could do that.

> :code:`exclude_rules = None` but :code:`exclude_rules != None`?

Reviewer: Or do you mean like this?:

```
Top-level folder .sqlfluffini: exclude_rules = None
      Sub-folder .sqlfluffini: exclude_rules != None
```

In that case the subfolder will override the top level folder config and for the queries in that solder folder they will exclude some rules. This can be handy if you know a folder can't handle a certain rule, but the rest of the project can.

Note also it's overridden completely as far as I know.

So, with this for example:

```
Top level folder .sqlfluffini: exclude_rules = L001`
      Sub Folder .sqlfluffini: exclude_rules = L002`
```

Queries in the sub folder would only exclude `L002` and would start to lint for `L001`. It would NOT exclude both because the closer config overrides, but is not cumulative.

> :code:`exclude_rules = None` but :code:`exclude_rules != None`?

Author: I replaced this paragraph with:

> If both :code:`exclude_rules` and :code:`rules` have non-empty value, then the
excluded rules are removed from the rules list. This allows for example
enabling common rules on top level but excluding some on subdirectory level.

I would say this is the reasonable level of detail. Thoughts?

> :code:`exclude_rules = None` but :code:`exclude_rules != None`?

Reviewer: That works for me!

Additionally, some rules have a special :code:`force_enable` configuration
option, which allows to enable the given rule even for dialects where it is
disabled by default. The rules that support this can be found in the
:ref:`ruleref`.

The default values can be seen in `Default Configuration`_.

See also: :ref:`inline_ignoring_errors` and `Ignoring Errors & Files`_.

> (**TODO** aren't those two pretty much overlapping?)

Author: This is strange. Is it ok to just link to both at this point?

Reviewer: Oh I don't think we should duplicate this content. One should just be a link to the other.

Not sure where is best lives? Under configuration or under rules? But since it's under rules already let's just add the link here and remove all the text.

Reviewer: Oh it's already duplicated in both! Never noticed that before!

OK let's leave as is. I think just the one link to the one on this page is fine. Might be worth doing a review to ensure they are both saying the same thing and sync up if not.

> 

Author: Should we also cross link back to the CLI args section from here?

Reviewer: I don't think you need to cross link here. We're not cross linking to the configuration file for `rules` and `exclude_rules` so why cross link to CLI which has the same?

Author: I was thinking, this has now become the main place to read about "how to enable or disable rules". For completeness, it could be worth pointing to the fact that rules can be also controlled via CLI args. That could be the best option in some situation. But of course it is almost obvious that user would check what `sqlfluff lint --help` has to offer. I don't have any real opinion on what should be done. Adding links everywhere can of lead into a bit bloated docs, yes. So we live without linking to CLI args from here?

Reviewer: Yeah I think leave it off here (for the bloated doc reason you give). It's not really the recommended way to ignore errors or include/exclude rules IMHO, but more a short hand for quite testing of only specific rules for that particular run.

of sqlfluff configuration. The value is a comma separated list of rule ids.

For example, to disable the rules :class:`L022 <sqlfluff.core.rules.Rule_L022>`
and :class:`L027 <sqlfluff.core.rules.Rule_L027>`:

.. code-block:: cfg

    [sqlfluff]
    exclude_rules = L022, L027

To enable individual rules, configure :code:`rules`, respectively.

For example, to enable :class:`L027 <sqlfluff.core.rules.Rule_L027>`:

.. code-block:: cfg

    [sqlfluff]
    rules = L027

If both :code:`exclude_rules` and :code:`rules` have non-empty value, then the
excluded rules are removed from the rules list. This allows for example
enabling common rules on top level but excluding some on subdirectory level.

Additionally, some rules have a special :code:`force_enable` configuration
option, which allows to enable the given rule even for dialects where it is
disabled by default. The rules that support this can be found in the
:ref:`ruleref`.

The default values can be seen in `Default Configuration`_.

See also: `Ignoring Errors & Files`_.

..
    TODO ^ There is also :ref:`inline_ignoring_errors` with overlapping
    content. Might be worth reviewing to ensure they are both saying the same

>     thing and sync up if not.

Author: I thought I'd leave this as a hidden comment:

<img width="719" alt="Näyttökuva 2022-1-19 kello 0 09 06" src="https://user-images.githubusercontent.com/4446608/150026657-2d52f2d2-add8-427d-b003-0c1041c77905.png">

Or it can be removed & create an issue about it?

Reviewer: That looks good to me. I say include it.
